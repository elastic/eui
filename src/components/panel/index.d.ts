/// <reference path="../common.d.ts" />

import { HTMLAttributes, SFC } from 'react';

declare module '@elastic/eui' {
  /**
   * panel type defs
   *
   * @see './panel.js'
   */

  export type PanelPaddingSize = 'none' | 's' | 'm' | 'l';

  export interface EuiPanelProps {
    hasShadow?: boolean;
    paddingSize?: PanelPaddingSize;
    grow?: boolean;
    panelRef?: RefCallback<HTMLDivElement>;
  }

  export const EuiPanel: SFC<
    CommonProps & HTMLAttributes<HTMLDivElement> & EuiPanelProps
  >;
}
