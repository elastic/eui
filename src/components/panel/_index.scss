@import 'panel';
@import './split_panel/index'
