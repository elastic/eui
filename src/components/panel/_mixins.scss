/**
 *  Mixin for use in:
 *  - EuiCard
*/
@mixin euiPanel($selector){
  @if variable-exists(selector) == false {
    @error "A $selector must be provided to @mixin euiPanel().";
  }
  @else {
    .#{$selector} {

      @include euiBottomShadowSmall;
      background-color: $euiColorEmptyShade;
      border: $euiBorderThin;
      border-radius: $euiBorderRadius;
      flex-grow: 1;
      &.#{$selector}--flexGrowZero {
        flex-grow: 0;
      }

      &.#{$selector}--isClickable {
        // in case of button wrapper which inherently is inline-block and no width
        display: block;
        width: 100%;
        text-align: left;

        // transition the shadow
        transition: all $euiAnimSpeedFast $euiAnimSlightResistance;

        &:hover,
        &:focus {
          @include euiSlightShadowHover;
          transform: translateY(-2px);
          cursor: pointer;
        }
      }

      &.#{$selector}--shadow {
        @if (lightness($euiTextColor) < 50) {
          border-bottom-color: rgba($euiShadowColor, .6);
        }
        @include euiBottomShadowMedium;

        &.#{$selector}--isClickable:hover,
        &.#{$selector}--isClickable:focus {
          @include euiBottomShadow;
        }
      }
    }
  }
}
