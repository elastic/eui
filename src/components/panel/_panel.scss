.euiPanel {
  @include euiBottomShadowSmall;

  background-color: $euiColorEmptyShade;
  border: $euiBorderThin;
  border-radius: $euiBorderRadius;
  flex-grow: 1;

  @each $modifier, $amount in $euiPanelPaddingModifiers {
    &.euiPanel--#{$modifier} {
      padding: $amount;
    }
  }

  &.euiPanel--shadow {
    @include euiBottomShadow;
  }

  &.euiPanel--flexGrowZero {
    flex-grow: 0;
  }
}
