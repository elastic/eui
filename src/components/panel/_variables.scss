/**
 * Padding map referenced in:
 *  - Popover
*/
$euiPanelPaddingModifiers: (
  "paddingSmall": $euiSizeS,
  "paddingMedium": $euiSize,
  "paddingLarge": $euiSizeL
) !default;