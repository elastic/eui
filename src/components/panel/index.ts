export { EuiPanel, PanelPaddingSize, SIZES } from './panel';
