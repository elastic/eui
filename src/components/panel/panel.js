import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const paddingSizeToClassNameMap = {
  none: null,
  s: 'euiPanel--paddingSmall',
  m: 'euiPanel--paddingMedium',
  l: 'euiPanel--paddingLarge',
};

export const SIZES = Object.keys(paddingSizeToClassNameMap);

export const EuiPanel = ({
  children,
  className,
  paddingSize,
  hasShadow,
  grow,
  panelRef,
  onClick,
  ...rest
}) => {

  const classes = classNames(
    'euiPanel',
    paddingSizeToClassNameMap[paddingSize],
    {
      'euiPanel--shadow': hasShadow,
      'euiPanel--flexGrowZero': !grow,
      'euiPanel--isClickable': onClick,
    },
    className
  );

  const PanelTag = onClick ? 'button' : 'div';

  return (
    <PanelTag
      onClick={onClick}
      ref={panelRef}
      className={classes}
      {...rest}
    >
      {children}
    </PanelTag>
  );

};

EuiPanel.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  hasShadow: PropTypes.bool,
  paddingSize: PropTypes.oneOf(SIZES),
  grow: PropTypes.bool,
  panelRef: PropTypes.func,
  onClick: PropTypes.func,
};

EuiPanel.defaultProps = {
  paddingSize: 'm',
  hasShadow: false,
  grow: true,
};
