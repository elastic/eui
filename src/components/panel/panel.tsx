/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, {
  ButtonHTMLAttributes,
  FunctionComponent,
  HTMLAttributes,
  Ref,
} from 'react';
import classNames from 'classnames';

import { CommonProps, keysOf, ExclusiveUnion } from '../common';

export const panelPaddingValues = {
  none: 0,
  s: 8,
  m: 16,
  l: 24,
};

const paddingSizeToClassNameMap = {
  none: null,
  s: 'euiPanel--paddingSmall',
  m: 'euiPanel--paddingMedium',
  l: 'euiPanel--paddingLarge',
};

export const SIZES = keysOf(paddingSizeToClassNameMap);

const borderRadiusToClassNameMap = {
  none: 'euiPanel--borderRadiusNone',
  m: 'euiPanel--borderRadiusMedium',
};

export const BORDER_RADII = keysOf(borderRadiusToClassNameMap);

export const COLORS = [
  'transparent',
  'plain',
  'subdued',
  'accent',
  'primary',
  'success',
  'warning',
  'danger',
] as const;

export type PanelColor = typeof COLORS[number];
export type PanelPaddingSize = typeof SIZES[number];
export type PanelBorderRadius = typeof BORDER_RADII[number];

export interface _EuiPanelProps extends CommonProps {
  /**
   * Adds a medium shadow to the panel;
   * Only works when `color="plain"`
   */
  hasShadow?: boolean;
  /**
   * Adds a slight 1px border on all edges.
   * Only works when `color="plain | transparent"`
   * Default is `undefined` and will default to that theme's panel style
   */
  hasBorder?: boolean;
  /**
   * Padding for all four sides
   */
  paddingSize?: PanelPaddingSize;
  /**
   * Corner border radius
   */
  borderRadius?: PanelBorderRadius;
  /**
   * When true the panel will grow in height to match `EuiFlexItem`
   */
  grow?: boolean;
  panelRef?: Ref<HTMLDivElement>;
  /**
   * Background color of the panel;
   * Usually a lightened form of the brand colors
   */
  color?: PanelColor;
}

export interface _EuiPanelDivlike
  extends _EuiPanelProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'color'> {
  element?: 'div';
}

export interface _EuiPanelButtonlike
  extends _EuiPanelProps,
    Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'color'> {
  element?: 'button';
}

export type EuiPanelProps = ExclusiveUnion<
  _EuiPanelButtonlike,
  _EuiPanelDivlike
>;

export const EuiPanel: FunctionComponent<EuiPanelProps> = ({
  children,
  className,
  paddingSize = 'm',
  borderRadius = 'm',
  color = 'plain',
  hasShadow = true,
  hasBorder,
  grow = true,
  panelRef,
  element,
  ...rest
}) => {
  // Shadows are only allowed when there's a white background (plain)
  const canHaveShadow = color === 'plain';
  const canHaveBorder = color === 'plain' || color === 'transparent';

  const classes = classNames(
    'euiPanel',
    paddingSizeToClassNameMap[paddingSize],
    borderRadiusToClassNameMap[borderRadius],
    `euiPanel--${color}`,
    {
      // The `no` classes turn off the option for default theme
      // While the `has` classes turn it on for Amsterdam
      'euiPanel--hasShadow': canHaveShadow && hasShadow === true,
      'euiPanel--noShadow': !canHaveShadow || hasShadow === false,
      'euiPanel--hasBorder': canHaveBorder && hasBorder === true,
      'euiPanel--noBorder': !canHaveBorder || hasBorder === false,
      'euiPanel--flexGrowZero': !grow,
      'euiPanel--isClickable': rest.onClick,
    },
    className
  );

  if (rest.onClick && element !== 'div') {
    return (
      <button
        ref={panelRef as Ref<HTMLButtonElement>}
        className={classes}
        {...(rest as ButtonHTMLAttributes<HTMLButtonElement>)}>
        {children}
      </button>
    );
  }

  return (
    <div
      ref={panelRef as Ref<HTMLDivElement>}
      className={classes}
      {...(rest as HTMLAttributes<HTMLDivElement>)}>
      {children}
    </div>
  );
};
