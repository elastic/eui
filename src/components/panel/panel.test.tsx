/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiPanel, SIZES, COLORS, BORDER_RADII } from './panel';

describe('EuiPanel', () => {
  test('is rendered', () => {
    const component = render(<EuiPanel {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('hasShadow', () => {
      test('can be false', () => {
        const component = render(<EuiPanel hasShadow={false} />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('grow', () => {
      test('can be false', () => {
        const component = render(<EuiPanel grow={false} />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('hasBorder', () => {
      test('can be false', () => {
        const component = render(<EuiPanel hasBorder={false} />);

        expect(component).toMatchSnapshot();
      });
      test('can be true', () => {
        const component = render(<EuiPanel hasBorder={true} />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('paddingSize', () => {
      SIZES.forEach((size) => {
        test(`${size} is rendered`, () => {
          const component = render(<EuiPanel paddingSize={size} />);

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('color', () => {
      COLORS.forEach((color) => {
        test(`${color} is rendered`, () => {
          const component = render(<EuiPanel color={color} />);

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('borderRadius', () => {
      BORDER_RADII.forEach((borderRadius) => {
        test(`${borderRadius} is rendered`, () => {
          const component = render(<EuiPanel borderRadius={borderRadius} />);

          expect(component).toMatchSnapshot();
        });
      });
    });
  });
});
