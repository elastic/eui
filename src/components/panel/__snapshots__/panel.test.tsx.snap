// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiPanel is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow testClass1 testClass2"
  data-test-subj="test subject string"
/>
`;

exports[`EuiPanel props borderRadius m is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props borderRadius none is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props color accent is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--accent euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props color danger is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--danger euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props color plain is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props color primary is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--primary euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props color subdued is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--subdued euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props color success is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--success euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props color transparent is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--transparent euiPanel--noShadow"
/>
`;

exports[`EuiPanel props color warning is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--warning euiPanel--noShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props grow can be false 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--flexGrowZero"
/>
`;

exports[`EuiPanel props hasBorder can be false 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--noBorder"
/>
`;

exports[`EuiPanel props hasBorder can be true 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder"
/>
`;

exports[`EuiPanel props hasShadow can be false 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow"
/>
`;

exports[`EuiPanel props paddingSize l is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingLarge euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props paddingSize m is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props paddingSize none is rendered 1`] = `
<div
  class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;

exports[`EuiPanel props paddingSize s is rendered 1`] = `
<div
  class="euiPanel euiPanel--paddingSmall euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
/>
`;
